front/src/App.test.js

```diff
@@ -43,12 +42,47 @@ describe("App.js", () => {
       expect(connectionCount).toBe(1);
       wrapper.unmount(this);
     }, 1000);
+    jest.runAllTimers();
     mockServer2.close();
   });
 
   it("renders without crashing", () => {
-    const div = document.createElement("div");
-    ReactDOM.render(<App />, div);
-    ReactDOM.unmountComponentAtNode(div);
+    shallow(<App />);
+  });
+
+  describe("updates the local state when it", () => {
+    let serverMessage, expectedState, initialTurtles;
+
+    afterEach(() => {
+      const mockServer = new Server(
+        "ws://user:testtoken@localhost/api/v1/state"
+      );
+      mockServer.on("connection", server => {
+        mockServer.send(serverMessage);
+      });
+      const wrapper = shallow(<App />);
+      wrapper.setState({ token: "testtoken", turtles: initialTurtles });
+      // Needed to make App.js connect to the mock server.
+      // If you use runAllTimers, it goes in a loop, if you run it <3 times, it doesn't work.
+      jest.runOnlyPendingTimers();
+      jest.runOnlyPendingTimers();
+      jest.runOnlyPendingTimers();
+      expect(JSON.stringify(wrapper.state("turtles"))).toBe(expectedState);
```

Comment:
Why not snapshot?

Reply:
This way we can define the result close to where the original object is defined. I could change it to snapshots though.